app/controllers/users/stock_roles.js

```diff
@@ -0,0 +1,189 @@
+    set(_, value) {
+      if (value === "accessTill") {
+        this.set("roleExpiryDate", moment().format("DD/MMM/YYYY"));
+      } else {
+        this.set("roleExpiryDate", "");
+      }
+
+      return value;
+    }
+  }),
+
+  observeExpiryDate: Ember.observer("roleExpiryDate", function() {
+    if (
+      this.get("roleExpiryDate") &&
+      this.get("stockRoleAccess") !== "accessTill"
+    ) {
+      this.set("stockRoleAccess", "accessTill");
+    }
+  }),
+
+  roleExpiryDate: Ember.computed("user.userRoles.[]", function() {
+    let date = this.get("userService").getRoleExpiryDate(
+      this.get("user"),
+      "stock"
+    );
+    return date ? moment(date).format("DD/MMM/YYYY") : "";
+  }),
+
+  hasStockAdministratorRole: Ember.computed("user.roles.[]", function() {
+    return this.get("userService").hasRole(
+      this.get("user"),
+      "Stock administrator"
+    );
+  }),
+
+  hasStockFulfilmentRole: Ember.computed("user.roles.[]", function() {
+    return this.get("userService").hasRole(
+      this.get("user"),
+      "Stock fulfilment"
+    );
+  }),
+
+  hasOrderAdministratorRole: Ember.computed("user.roles.[]", function() {
+    return this.get("userService").hasRole(
+      this.get("user"),
+      "Order administrator"
+    );
+  }),
+
+  hasOrderFulfilmentRole: Ember.computed("user.roles.[]", function() {
+    return this.get("userService").hasRole(
+      this.get("user"),
+      "Order fulfilment"
+    );
+  }),
+
+  noStockAppRole: Ember.computed(
+    "hasStockAdministratorRole",
+    "hasStockFulfilmentRole",
+    "hasOrderAdministratorRole",
+    "hasOrderFulfilmentRole",
+    function() {
+      return (
+        !this.get("hasStockAdministratorRole") &&
+        !this.get("hasStockFulfilmentRole") &&
+        !this.get("hasOrderAdministratorRole") &&
+        !this.get("hasOrderFulfilmentRole")
+      );
+    }
+  ),
+
+  updateUserRole(role, selectedRole, roleExpiryDate) {
+    const roleId = this.get("userService").getRoleId(role);
+    const userId = this.get("user.id");
+
+    if (selectedRole) {
+      this.get("userService").assignRole(userId, roleId, roleExpiryDate);
+    } else {
+      this.get("userService").deleteUserRole(userId, roleId);
+    }
+  },
+
+  actions: {
+    setPrinterValue(value) {
+      const printerId = value.id;
+      this.set("selectedPrinterId", printerId);
+    },
+
+    cancelForm() {
+      this.set("selectedPrinterId", "");
+      this.transitionToRoute("users.details", this.get("user.id"));
+    },
+
+    saveUserRoles() {
+      let roleExpiryDate;
+      let stockRoleAccess = this.get("stockRoleAccess");
+      const userId = this.get("user.id");
+
+      if (stockRoleAccess === "noAccess") {
+        this.get("userService").deleteStockRoles(this.get("user"));
+      } else {
+        if (this.get("canUpdateRole")) {
+          if (stockRoleAccess === "accessTill") {
+            roleExpiryDate = this.get("roleExpiryDate");
+          }
+
+          this.updateUserRole(
```

Comment:
I think here you should send array of `permission` to the service and there you can iterate and update with other helper method.

Reply:
Hardcoding the role names definitely makes my spidey sense tingle. This should be derived from the list of roles you fetch from the API. 

If I add a new role on the backend, we should not have to update the UI at all to support it